src/DotNetLightning.Core/Chain/KeysInterface.fs

```diff
@@ -47,146 +47,3 @@ type SpendableOutputDescriptor =
     | DynamicOutputP2WSH of DynamicOutputP2WSH
     | DynamicOutputP2WPKH of DynamicOutputP2WPKH
 
-/// In usual operation we should not hold secrets on memory. So only hold pubkey
-type ChannelPubKeys = {
-    FundingPubKey: PubKey
-    RevocationBasePubKey: PubKey
-    PaymentBasePubKey: PubKey
-    DelayedPaymentBasePubKey: PubKey
-    HTLCBasePubKey: PubKey
-}
-
-/// Set of lightning keys needed to operate a channel as describe in BOLT 3
-type ChannelKeys = {
-    FundingKey: Key
-    RevocationBaseKey: Key
-    PaymentBaseKey: Key
-    DelayedPaymentBaseKey: Key
-    HTLCBaseKey: Key
-    CommitmentSeed: CommitmentSeed
-}
-
-    with
-        member this.ToChannelPubKeys() =
-            {
-                FundingPubKey = this.FundingKey.PubKey
-                RevocationBasePubKey = this.RevocationBaseKey.PubKey
-                PaymentBasePubKey = this.PaymentBaseKey.PubKey
-                DelayedPaymentBasePubKey = this.DelayedPaymentBaseKey.PubKey
-                HTLCBasePubKey = this.HTLCBaseKey.PubKey
-            }
-
-
-/// Interface to describe an object which can get user secrets and key material.
-type IKeysRepository =
```

Comment:
I don't understand what you mean. As it was, `ChannelPrivKeys` and `IKeysRepository` were doing pretty much the exact same thing. So I can't see the point in having both.

Reply:
I think I get what you're saying now. Correct me if I'm wrong...

Holding private keys in RAM is dangerous because another process could read our memory and get the keys. `ChannelPrivKeys` is no different from `DefaultKeyRepository` in this regard (since they're pretty much the same type) however the keys repository was designed as an interface so that a user could replace it with something more secure.

I think if we want to take that design seriously then there's some things to consider:

Firstly, `IKeysRepository` shouldn't have an option to get the channel private keys at all, otherwise we'll end up holding them in RAM anyway. Even if the code only grabs the keys temporarily when it needs them we don't have a way to control when those keys get garbage collected. Also they'll still live on in RAM for a while after they get collected, assuming the CLR doesn't zero-out memory once it gets collected since that would be an expensive waste of time when the memory is being used for anything other than holding secrets.

Also, the user can't define an `IKeysRepository` that requires user intervention whenever DotNetLightning needs to access the keys. This is particularly true for the `NodeSecret` since we need it every time we so much as reply to a ping. But it's also true of the channel keys since you need them to create commitment transactions, unless (maybe?) if other nodes are willing to deal with a node which can't forward htlcs and which is extremely unresponsive when accepting payments.

We only actually need to hold the `NodeMasterKey`, rather than a hash map of private keys, since all the other private keys are derived from it.

So maybe a good design here would be to (a) reinstate `IKeysRepository` as a replacement for `NodeMasterKey`. (b) Give it some very-generic encrypt and sign methods which take as an argument the derivation path to the derived key to use. (c) Replace all other instances of `Key` in the code with a new type which wraps a reference to an `IKeysRepository` and a derivation path.

That way no secret data would ever leave the `IKeysRepository` type and a type which implements the `IKeysRepository` interface could store the key in mlocked memory or something.

Does that sound like a good way forward?

Reply:
I think you understand what I meant perfectly (Indeed better than me.)

> Firstly, IKeysRepository shouldn't have an option to get the channel private keys at all, otherwise we'll end up holding them in RAM anyway. Even if the code only grabs the keys temporarily when it needs them we don't have a way to control when those keys get garbage collected. Also they'll still live on in RAM for a while after they get collected, assuming the CLR doesn't zero-out memory once it gets collected since that would be an expensive waste of time when the memory is being used for anything other than holding secrets.

Yeah this is very good point. I had to make `ChannelKeys` a `ref struct`. I actually tried once, but it does not go well with functional style. computation expression did not work with ref struct.

> So maybe a good design here would be to (a) reinstate IKeysRepository as a replacement for NodeMasterKey. (b) Give it some very-generic encrypt and sign methods which take as an argument the derivation path to the derived key to use. (c) Replace all other instances of Key in the code with a new type which wraps a reference to an IKeysRepository and a derivation path.

This sounds good. But I think the refactoring will be very hard. Unless you want to try, I think we should go as is so that we can first get shit done.

So yeah, I'm not ok with removing the `IKeysRepository` there are tons of other things we have to consider in terms of security anyway...

Reply:
@canndrew please note that it seems Joe edited his last comment to change `not` to `now` hehe ;-)